chain/network/src/concurrency/scope/mod.rs

```diff
@@ -0,0 +1,459 @@
+            } else {
+                Ok(())
+            }
+        }
+    }
+
+    /// Spawns a task in this scope.
+    ///
+    /// Returns ErrTerminated if the scope has already terminated.
+    pub fn spawn<T: 'static + Send>(
+        &self,
+        f: impl 'static + Send + Future<Output = Result<T, E>>,
+    ) -> Result<(), ErrTerminated> {
+        self.0
+            .upgrade()
+            .map(|m| {
+                Inner::spawn(m, f);
+            })
+            .ok_or(ErrTerminated)
+    }
+
+    /// Spawns a service in this scope.
+    ///
+    /// Returns ErrTerminated if the scope has already terminated.
+    pub fn new_service(&self) -> Result<Service<E>, ErrTerminated> {
+        self.0.upgrade().map(|m| Inner::new_service(m)).ok_or(ErrTerminated)
+    }
+}
+
+/// Wrapper of a scope reference which cancels the scope when dropped.
+///
+/// Used by Scope to cancel the scope as soon as all tasks spawned via
+/// `Scope::spawn` complete.
+struct StrongService<E: 'static>(Arc<Inner<E>>);
+
+impl<E: 'static> Borrow<Inner<E>> for StrongService<E> {
+    fn borrow(&self) -> &Inner<E> {
+        &*self.0
+    }
+}
+
+impl<E: 'static> Drop for StrongService<E> {
+    fn drop(&mut self) {
+        self.0.ctx.cancel()
+    }
+}
+
+/// Scope represents a concurrent computation bounded by lifetime 'env.
+///
+/// It should be created only via `run!` macro.
+/// Scope is cancelled when the provided context is cancelled.
+/// Scope is cancelled when any of the tasks in the scope returns an error.
+/// Scope is cancelled when all the tasks in the scope complete.
+/// Scope is terminated when it is cancelled AND all tasks in the scope complete.
+pub struct Scope<'env, E: 'static>(
+    /// Scope is equivalent to a strong service, but bounds
+    Weak<StrongService<E>>,
+    Weak<Inner<E>>,
+    /// Makes Scope<'env,E> invariant in 'env.
+    std::marker::PhantomData<fn(&'env ()) -> &'env ()>,
+);
+
+unsafe fn to_static<'env, T>(f: BoxFuture<'env, T>) -> BoxFuture<'static, T> {
+    std::mem::transmute::<BoxFuture<'env, _>, BoxFuture<'static, _>>(f)
+}
+
+impl<'env, E: 'static + Send> Scope<'env, E> {
+    /// Spawns a "main" task in the scope.
+    /// Scope gets canceled as soon as all the "main" tasks complete.
+    pub fn spawn<T: 'static + Send>(
+        &self,
+        f: impl 'env + Send + Future<Output = Result<T, E>>,
+    ) -> JoinHandle<'env, T> {
+        match self.0.upgrade() {
+            Some(strong) => JoinHandle(
+                Inner::spawn(strong, unsafe { to_static(f.boxed()) }),
+                std::marker::PhantomData,
+            ),
+            // Upgrade may fail only if all the "main" tasks have already completed
+            // so the caller is a "background" task. In that case we fall back
+            // to spawning a "background" task instead. It is ok, since the distinction
+            // between main task and background task disappears, once the scope is canceled.
+            None => self.spawn_bg(f),
```

Comment:
If the scope gets cancelled as soon as no main tasks are left, why even bother with spawning it as a bg task? As AFAIU it should stop before the first `await` point, it can’t do anything meaningful anyway.

Reply:
A background task doesn't have to have any await point in principle. And even if it did, we allow it to get canceled on its own. Essentially tasks spawned via spawn_bg should only spawn bg tasks, however it is hard to prevent them from calling spawn() as it has access to &Scope. Instead, we just allow it to call spawn(), which just won't be able to "uncancel" the canceled scope. It is ok, because a scope can be cancelled from the outside of the scope, and it doesn't change the fact that the main tasks can still spawn other main tasks.